dm/worker/upgrade.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// MarshalBinary implements encoding.BinaryMarshal.
+func (v *version) MarshalBinary() ([]byte, error) {
+	return json.Marshal(v)
+}
+
+// UnmarshalBinary implements encoding.BinaryMarshal.
+func (v *version) UnmarshalBinary(data []byte) error {
+	return json.Unmarshal(data, v)
+}
+
+// loadVersion loads the version of DM-worker from the levelDB.
+func loadVersion(h Getter) (ver version, err error) {
+	if whetherNil(h) {
+		return ver, errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := h.Get(dmWorkerVersionKey, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return defaultPreviousWorkerVersion, nil
+		}
+		return ver, errors.Annotatef(err, "load version with key %s from levelDB", dmWorkerVersionKey)
+	}
+	err = ver.UnmarshalBinary(data)
+	return ver, errors.Annotatef(err, "unmarshal version from data % X", data)
+}
+
+// saveVersion saves the version of DM-worker into the levelDB.
+func saveVersion(h Putter, ver version) error {
+	if whetherNil(h) {
+		return errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := ver.MarshalBinary()
+	if err != nil {
+		return errors.Annotatef(err, "marshal version %v to binary data", ver)
+	}
+
+	err = h.Put(dmWorkerVersionKey, data, nil)
+	return errors.Annotatef(err, "save version %v into levelDB with key %s", ver, dmWorkerVersionKey)
+}
+
+// tryUpgrade tries to upgrade from an older version.
+func tryUpgrade(dbDir string) error {
+	// 1. check the DB directory
+	notExist := false
+	fs, err := os.Stat(dbDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			notExist = true
+		} else {
+			return errors.Annotatef(err, "get stat for %s", dbDir)
+		}
+	} else if !fs.IsDir() { // should be a directory
+		return errors.NotValidf("directory %s for DB", dbDir)
+	}
+
+	// 2. open the kv DB
+	db, err := openDB(dbDir, defaultKVConfig)
+	if err != nil {
+		return errors.Annotatef(err, "open DB for %s", dbDir)
+	}
+	defer func() {
+		err = db.Close()
+		if err != nil {
+			log.Errorf("[worker upgrade] close DB fail %v", err)
+		}
+	}()
+
+	if notExist {
+		log.Infof("[worker upgrade] no previous operation log exists, no need to upgrade")
+		// still need to save the current version version
+		currVer := newCurrentVersion()
+		err = saveVersion(db, currVer)
+		return errors.Annotatef(err, "save current version %v into DB %s", currVer, dbDir)
+	}
+
+	// 3. load previous version
+	prevVer, err := loadVersion(db)
+	if err != nil {
+		return errors.Annotatef(err, "load previous version from DB %s", dbDir)
+	}
+	log.Infof("[worker upgrade] the previous version is %v", prevVer)
+
+	// 4. check needing to upgrade
+	currVer := newCurrentVersion()
+	if prevVer.compare(currVer) == 0 {
+		log.Infof("[worker upgrade] the previous and current versions both are %v, no need to upgrade", prevVer)
+		return nil
+	} else if prevVer.compare(currVer) > 0 {
+		log.Warnf("[worker upgrade] the previous version %v is newer than current %v, automatic downgrade is not supported now", prevVer, currVer)
+		return nil
```

Comment:
is it better to return error here?

Reply:
changed in https://github.com/pingcap/dm/pull/187/commits/10281b801665c53de0201cd2cb9a1a82dc2ccec7.